File: backend/shared/src/domain/entities/ehr/lab_result.rs
//! EHR Lab Result Entity
//!
//! Corresponds to VistA File #63 (^LR) - Lab Results

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use uuid::Uuid;

use crate::shared::AuditFields;

/// Lab result status
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "VARCHAR")]
#[sqlx(rename_all = "lowercase")]
pub enum LabStatus {
    #[serde(rename = "pending")]
    Pending,
    #[serde(rename = "preliminary")]
    Preliminary,
    #[serde(rename = "final")]
    Final,
    #[serde(rename = "corrected")]
    Corrected,
    #[serde(rename = "cancelled")]
    Cancelled,
}

impl Default for LabStatus {
    fn default() -> Self {
        LabStatus::Pending
    }
}

/// Abnormality flag
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "VARCHAR")]
#[sqlx(rename_all = "lowercase")]
pub enum AbnormalFlag {
    #[serde(rename = "normal")]
    Normal,
    #[serde(rename = "low")]
    Low,
    #[serde(rename = "high")]
    High,
    #[serde(rename = "critical_low")]
    CriticalLow,
    #[serde(rename = "critical_high")]
    CriticalHigh,
    #[serde(rename = "abnormal")]
    Abnormal,
}

impl Default for AbnormalFlag {
    fn default() -> Self {
        AbnormalFlag::Normal
    }
}

/// EHR Lab Result Entity
///
/// Represents a laboratory test result.
/// Corresponds to VistA File #63 (^LR - Lab Results).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EhrLabResult {
    pub id: Uuid,

    /// Internal Entry Number (VistA IEN)
    pub ien: i64,

    /// Organization (multi-tenant)
    pub organization_id: Uuid,

    /// Patient ID
    pub patient_id: Uuid,

    /// Visit when ordered
    pub visit_id: Option<Uuid>,

    /// Order ID (if from order entry)
    pub order_id: Option<Uuid>,

    /// LOINC code
    pub loinc_code: Option<String>,

    /// Test name
    pub test_name: String,

    /// Test category (chemistry, hematology, etc.)
    pub category: Option<String>,

    /// Result value (as string for flexibility)
    pub result_value: Option<String>,

    /// Numeric value (if applicable)
    pub numeric_value: Option<f64>,

    /// Unit of measure
    pub unit: Option<String>,

    /// Reference range (e.g., "70-100")
    pub reference_range: Option<String>,

    /// Reference range low
    pub reference_low: Option<f64>,

    /// Reference range high
    pub reference_high: Option<f64>,

    /// Status
    pub status: LabStatus,

    /// Abnormality flag
    pub abnormal_flag: AbnormalFlag,

    /// Specimen type (blood, urine, etc.)
    pub specimen_type: Option<String>,

    /// Collection datetime
    pub collection_datetime: Option<DateTime<Utc>>,

    /// Received datetime
    pub received_datetime: Option<DateTime<Utc>>,

    /// Result datetime
    pub result_datetime: Option<DateTime<Utc>>,

    /// Performing lab
    pub performing_lab: Option<String>,

    /// Ordering provider
    pub ordering_provider_id: Option<Uuid>,

    /// Comments
    pub comments: Option<String>,

    /// MUMPS global data
    pub mumps_data: Option<JsonValue>,

    // Audit fields
    pub request_id: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub created_by: Option<Uuid>,
    pub updated_by: Option<Uuid>,
    pub system_id: Option<String>,
    pub version: i64,
}

impl EhrLabResult {
    /// Create a new lab result
    pub fn new(
        organization_id: Uuid,
        patient_id: Uuid,
        test_name: String,
    ) -> Self {
        let audit = AuditFields::new();
        Self {
            id: Uuid::new_v4(),
            ien: 0,
            organization_id,
            patient_id,
            visit_id: None,
            order_id: None,
            loinc_code: None,
            test_name,
            category: None,
            result_value: None,
            numeric_value: None,
            unit: None,
            reference_range: None,
            reference_low: None,
            reference_high: None,
            status: LabStatus::Pending,
            abnormal_flag: AbnormalFlag::Normal,
            specimen_type: None,
            collection_datetime: None,
            received_datetime: None,
            result_datetime: None,
            performing_lab: None,
            ordering_provider_id: None,
            comments: None,
            mumps_data: None,
            request_id: audit.request_id,
            created_at: audit.created_at,
            updated_at: audit.updated_at,
            created_by: audit.created_by,
            updated_by: audit.updated_by,
            system_id: audit.system_id,
            version: audit.version,
        }
    }

    /// Set the result with automatic abnormal flag detection
    pub fn set_result(&mut self, value: f64, unit: String) {
        self.numeric_value = Some(value);
        self.result_value = Some(value.to_string());
        self.unit = Some(unit);
        self.result_datetime = Some(Utc::now());
        self.status = LabStatus::Final;

        // Determine abnormal flag if reference range is set
        self.abnormal_flag = self.calculate_abnormal_flag(value);
        self.updated_at = Utc::now();
    }

    /// Calculate abnormal flag based on reference range
    fn calculate_abnormal_flag(&self, value: f64) -> AbnormalFlag {
        match (self.reference_low, self.reference_high) {
            (Some(low), Some(high)) => {
                // Critical ranges (typically 1.5x the normal range)
                let critical_low = low - (high - low) * 0.5;
                let critical_high = high + (high - low) * 0.5;

                if value < critical_low {
                    AbnormalFlag::CriticalLow
                } else if value > critical_high {
                    AbnormalFlag::CriticalHigh
                } else if value < low {
                    AbnormalFlag::Low
                } else if value > high {
                    AbnormalFlag::High
                } else {
                    AbnormalFlag::Normal
                }
            }
            (Some(low), None) => {
                if value < low {
                    AbnormalFlag::Low
                } else {
                    AbnormalFlag::Normal
                }
            }
            (None, Some(high)) => {
                if value > high {
                    AbnormalFlag::High
                } else {
                    AbnormalFlag::Normal
                }
            }
            _ => AbnormalFlag::Normal,
        }
    }

    /// Check if result is critical
    pub fn is_critical(&self) -> bool {
        matches!(
            self.abnormal_flag,
            AbnormalFlag::CriticalLow | AbnormalFlag::CriticalHigh
        )
    }

    /// Check if result is abnormal
    pub fn is_abnormal(&self) -> bool {
        !matches!(self.abnormal_flag, AbnormalFlag::Normal)
    }

    /// Get display value with unit
    pub fn display_value(&self) -> String {
        let value = self.result_value.as_deref().unwrap_or("-");
        if let Some(ref unit) = self.unit {
            format!("{} {}", value, unit)
        } else {
            value.to_string()
        }
    }

    /// Get display with reference range
    pub fn display_with_range(&self) -> String {
        let mut display = self.display_value();
        if let Some(ref range) = self.reference_range {
            display.push_str(&format!(" (ref: {})", range));
        }
        display
    }
}

File: cli/packages/libs/shared/src/schemas/ehr/lab-result.ts
/**
 * EHR Lab Result Schemas
 *
 * Runtime validation for laboratory test results.
 * Corresponds to VistA File #63 (^LR)
 * Single source of truth - use z.infer<> for types.
 */

import { z } from "zod";
import { DateTimeSchema } from "../common";
import { createTypeGuard, createAssertion } from "../guards";

// ============================================================================
// Lab Enums
// ============================================================================

/**
 * Lab result status schema
 */
export const EhrLabStatusSchema = z.enum([
  "pending",
  "in_progress",
  "completed",
  "cancelled",
]);

export type EhrLabStatus = z.infer<typeof EhrLabStatusSchema>;

/**
 * Abnormality flag schema
 */
export const EhrAbnormalFlagSchema = z.enum([
  "normal",
  "low",
  "high",
  "critical_low",
  "critical_high",
  "abnormal",
]);

export type EhrAbnormalFlag = z.infer<typeof EhrAbnormalFlagSchema>;

// ============================================================================
// Lab Result Schema
// ============================================================================

/**
 * Full EHR Lab Result schema
 */
export const EhrLabResultSchema = z.object({
  // Identity
  id: z.string().uuid(),
  ien: z.number().int().positive(),
  organizationId: z.string().uuid(),
  patientId: z.string().uuid(),
  visitId: z.string().uuid().optional(),
  orderId: z.string().uuid().optional(),

  // Test identification
  loincCode: z.string().optional(),
  testName: z.string().min(1, { message: "Test name is required" }).max(300),
  category: z.string().max(100).optional(),

  // Result values
  value: z.string().max(500),
  numericValue: z.number().optional(),
  unit: z.string().max(50).optional(),

  // Reference ranges
  referenceRange: z.string().max(100).optional(),
  referenceLow: z.number().optional(),
  referenceHigh: z.number().optional(),

  // Status and flags
  status: EhrLabStatusSchema,
  abnormalFlag: EhrAbnormalFlagSchema.optional(),

  // Specimen and timing
  specimenType: z.string().max(100).optional(),
  collectedAt: DateTimeSchema.optional(),
  receivedAt: DateTimeSchema.optional(),
  resultedAt: DateTimeSchema.optional(),

  // Provider and interpretation
  performingLab: z.string().max(200).optional(),
  orderingProviderId: z.string().uuid().optional(),
  interpretation: z.string().max(2000).optional(),
  notes: z.string().max(2000).optional(),

  // Audit fields
  createdAt: DateTimeSchema.optional(),
  createdBy: z.string().optional(),
  updatedAt: DateTimeSchema.optional(),
  updatedBy: z.string().optional(),

  // MUMPS data
  mumpsData: z.record(z.unknown()).optional(),
}).refine(
  // Business rule: If numeric value exists and reference ranges exist, calculate abnormal flag
  data => {
    if (data.numericValue !== undefined && data.referenceLow !== undefined && data.referenceHigh !== undefined) {
      // This is just a validation check; actual calculation should be done server-side
      return true;
    }
    return true;
  },
  {
    message: "Numeric results with reference ranges should have abnormal flag calculated",
    path: ["abnormalFlag"],
  }
);

export type EhrLabResult = z.infer<typeof EhrLabResultSchema>;

// Type guards
export const isEhrLabResult = createTypeGuard(EhrLabResultSchema);

// Assertions
export const assertEhrLabResult = createAssertion(EhrLabResultSchema, 'EhrLabResult');

// ============================================================================
// Create Lab Result Request Schema
// ============================================================================

export const CreateEhrLabResultRequestSchema = z.object({
  patientId: z.string().uuid(),
  visitId: z.string().uuid().optional(),
  orderId: z.string().uuid().optional(),
  loincCode: z.string().optional(),
  testName: z.string().min(1).max(300),
  category: z.string().max(100).optional(),
  resultValue: z.string().max(500).optional(),
  numericValue: z.number().optional(),
  unit: z.string().max(50).optional(),
  referenceRange: z.string().max(100).optional(),
  referenceLow: z.number().optional(),
  referenceHigh: z.number().optional(),
  specimenType: z.string().max(100).optional(),
  collectionDatetime: DateTimeSchema.optional(),
  comments: z.string().max(2000).optional(),
});

export type CreateEhrLabResultRequest = z.infer<typeof CreateEhrLabResultRequestSchema>;

export const isCreateEhrLabResultRequest = createTypeGuard(CreateEhrLabResultRequestSchema);
export const assertCreateEhrLabResultRequest = createAssertion(CreateEhrLabResultRequestSchema, 'CreateEhrLabResultRequest');

// ============================================================================
// Update Lab Result Request Schema
// ============================================================================

export const UpdateEhrLabResultRequestSchema = CreateEhrLabResultRequestSchema.partial().extend({
  id: z.string().uuid(),
  status: EhrLabStatusSchema.optional(),
});

export type UpdateEhrLabResultRequest = z.infer<typeof UpdateEhrLabResultRequestSchema>;

export const isUpdateEhrLabResultRequest = createTypeGuard(UpdateEhrLabResultRequestSchema);
export const assertUpdateEhrLabResultRequest = createAssertion(UpdateEhrLabResultRequestSchema, 'UpdateEhrLabResultRequest');

// ============================================================================
// Lab Search Criteria Schema
// ============================================================================

export const EhrLabSearchCriteriaSchema = z.object({
  patientId: z.string().uuid().optional(),
  visitId: z.string().uuid().optional(),
  orderId: z.string().uuid().optional(),
  status: EhrLabStatusSchema.optional(),
  testName: z.string().optional(),
  loincCode: z.string().optional(),
  category: z.string().optional(),
  abnormalFlag: EhrAbnormalFlagSchema.optional(),
  dateFrom: DateTimeSchema.optional(),
  dateTo: DateTimeSchema.optional(),
  abnormalOnly: z.boolean().optional(),
  criticalOnly: z.boolean().optional(),
});

export type EhrLabSearchCriteria = z.infer<typeof EhrLabSearchCriteriaSchema>;

export const isEhrLabSearchCriteria = createTypeGuard(EhrLabSearchCriteriaSchema);

// ============================================================================
// Lab Panel Schema
// ============================================================================

/**
 * Lab panel (group of related tests) schema
 */
export const EhrLabPanelSchema = z.object({
  panelName: z.string(),
  loincCode: z.string(),
  results: z.array(EhrLabResultSchema),
  collectionDatetime: DateTimeSchema,
});

export type EhrLabPanel = z.infer<typeof EhrLabPanelSchema>;

export const isEhrLabPanel = createTypeGuard(EhrLabPanelSchema);

// ============================================================================
// Utility Functions
// ============================================================================

export function isLabCompleted(lab: EhrLabResult): boolean {
  return lab.status === "completed";
}

export function isLabAbnormal(lab: EhrLabResult): boolean {
  return lab.abnormalFlag !== undefined && lab.abnormalFlag !== "normal";
}

export function isLabCritical(lab: EhrLabResult): boolean {
  return lab.abnormalFlag === "critical_low" || lab.abnormalFlag === "critical_high";
}

export function calculateAbnormalFlag(
  numericValue: number,
  referenceLow?: number,
  referenceHigh?: number
): EhrAbnormalFlag {
  if (referenceLow === undefined || referenceHigh === undefined) {
    return "normal";
  }

  const criticalLowThreshold = referenceLow * 0.5;
  const criticalHighThreshold = referenceHigh * 1.5;

  if (numericValue < criticalLowThreshold) {
    return "critical_low";
  }
  if (numericValue > criticalHighThreshold) {
    return "critical_high";
  }
  if (numericValue < referenceLow) {
    return "low";
  }
  if (numericValue > referenceHigh) {
    return "high";
  }

  return "normal";
}
